### PR TITLE
control-plane hardening: cleanup local state on peer leaving a network

### PR DESCRIPTION
**Pull request by sanimej**
When the last task on an overlay network is shutdown, the node will leave from the network. Once a node leaves the network subsequence bulksync updates won't include the state for that network. If a bunch of tasks on a network are shutdown when there other network issues some of the gossip updates might be lost. If the network leave event does reach the peers they should cleanup the local state. Not doing so will leave the missed gossip state updates permanently in the peers' networkdb.

Signed-off-by: Santhosh Manohar <santhosh@docker.com>

### CODE REVIEW

File: networkdb/networkdb.go [hunk 1]
```diff
@@ -372,6 +372,37 @@ func (nDB *NetworkDB) deleteNetworkEntriesForNode(deletedNode string) {
 	nDB.Unlock()
 }
 
+func (nDB *NetworkDB) deleteNodeNetworkEntries(nid, node string) {
+	nDB.Lock()
```
Comment (aboch) [inline networkdb/networkdb.go]: Is it necessary or can we avoid to to lock the nDB around the walk ?

Comment (sanimej) [inline networkdb/networkdb.go]: @aboch We have to take the lock in this case since the network DB entries have to updated.

**Comment by mavenugo** [comment]
LGTM

**Comment by aboch** [comment]
LGTM

**Pull request merged**